### ISSUE TITLE
Dialyzer does not accept structs created with `new/0`, `new/1`... 

### ISSUE THREAD

**Issue by danmarcab** [elixir-protobuf/protobuf#287]
If we have the following code:

```elixir
  alias Google.Protobuf.Timestamp

  @spec fine :: Timestamp.t()
  def fine do
    %Timestamp{seconds: 232_342, nanos: 232_342}
  end

  @spec not_fine :: Timestamp.t()
  def not_fine do
    Timestamp.new(seconds: 232_342, nanos: 232_342)
  end
```

The first one is ok according to dialyzer. The second one isn't. 
It seems the type spec of the behaviour https://github.com/elixir-protobuf/protobuf/blob/main/lib/protobuf.ex#L81 is `struct()` and dialyzer doesn't accept any struct as the concrete type.

I created a repo with a minimal case: 
https://github.com/danmarcab/protobuf_dialyzer_error_test_repo

**Comment by danmarcab** [comment]
Sure, for the above code you get:

```
Total errors: 1, Skipped: 0, Unnecessary Skips: 0
done in 0m0.9s
lib/protobuf_dialyzer_error_test_repo.ex:26:invalid_contract
The @spec for the function does not match the success typing of the function.

Function:
ProtobufDialyzerErrorTestRepo.not_fine/0

Success typing:
@spec not_fine() :: %{:__struct__ => atom()}

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

**Comment by whatyouhide** [comment]
Closed via #288.
